# **Data Cleaning Main Datavista and Area Dataset**

This is the initial stage of developing a room pricing model, where data reading and cleaning will be carried out from the main datavista and also the area dataset. First we can import all used packages and libraries

In [44]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from tabulate import tabulate
import seaborn as sns
import warnings

In [26]:
warnings.filterwarnings("ignore")
warnings.filterwarnings("ignore",category=FutureWarning)

### **Data Wrangling**

**Gathering Data**



In this stage, the datasets, namely datavista and area are collected and then we will read the datasets using pandas library.

In [27]:
main_df = pd.read_csv("main.csv", delimiter=",")
main_df.head()

property_id unit_id  average_baseline_price  bedroom  bathroom  beds  \
0      PR0017  UN0026            3.825851e+05        1       1.0   1.0   
1      PR0045  UN1518            4.847391e+05        1       1.0   1.0   
2      PR0051  UN1471            7.079566e+05        3       2.0   3.0   
3      PR0055  UN1417            7.403196e+05        3       2.0   3.0   
4      PR0065  UN1392            2.154133e+06        8       8.0   8.0   

   capacity   ac  balcony  beachfront  ...  wifi  workspace       lat  \
0       2.0  1.0      1.0         0.0  ...   1.0        1.0 -8.656211   
1       2.0  1.0      0.0         0.0  ...   0.0        1.0 -8.698719   
2       6.0  1.0      0.0         0.0  ...   1.0        1.0       NaN   
3       7.0  1.0      1.0         0.0  ...   0.0        1.0 -7.831766   
4       6.0  1.0      0.0         0.0  ...   1.0        1.0       NaN   

          lng         property_name  \
0  115.132867  Double D Guest House   
1  115.163846             Melin Inn   
2         NaN           Omah JeTeHa   
3  110.380636       Griya Wirosaban   
4         NaN            Pendopo 26   

                                  property_direction  property_type  \
0  Double''D Guest House, Jalan Nelayan, Canggu, ...            4.0   
1          Melin Inn, Legian, Kabupaten Badung, Bali            3.0   
2  Jl. Sorowajan Baru, perum Jogja Town House 3 n...            1.0   
3  Jl. Ki Ageng Pemanahan, Kragilan, Tamanan, Kec...            1.0   
4                                                NaN            2.0   

   property_design  area_id  distance_to_coastline  
0              0.0    A0003                    NaN  
1              0.0    A0015                    NaN  
2              0.0    A0044                    NaN  
3              0.0    A0001                    NaN  
4              0.0    A0001                    NaN  

[5 rows x 40 columns]

In [28]:
main_df.shape

(357, 40)

After reading the dataset, it is found that this data has a total of 40 features and 357 rows. Then, in the same way, we will also apply this to the arecodes dataset.

In [29]:
areacodes_df = pd.read_csv("areacodes.csv", delimiter=",")
areacodes_df.head()

area_id   area_name  area_squad employee_id created_at        updated_at  \
0   A0000   Undefined         NaN         NaN        NaN               NaN   
1   A0001  Yogyakarta         3.0       E0028        NaN  2023-05-17 04:14   
2   A0002     Uluwatu         2.0       E0015        NaN  2022-10-07 06:34   
3   A0003      Canggu         2.0       E0014        NaN  2022-10-07 06:34   
4   A0004        Ubud         2.0       E0042        NaN  2022-10-07 06:34   

   deleted_at  temp_column area_distance_to_airport  airport_pickup_price_idr  \
0         NaN          NaN                      NaN                       NaN   
1         NaN          NaN               30 minutes                       NaN   
2         NaN          NaN                   1 hour                  325000.0   
3         NaN          NaN                   1 hour                  325000.0   
4         NaN          NaN                  2 hours                  585000.0   

   area_airport_name  region_id  
0                NaN        NaN  
1  Adi Sucipto (JOG)        2.0  
2   Ngurah Rai (DPS)        1.0  
3   Ngurah Rai (DPS)        1.0  
4   Ngurah Rai (DPS)        1.0

In [30]:
areacodes_df.shape

(54, 12)

It can be seen that the areacodes dataset has 54 area IDs and also has 12 features.

### **Assessing Data**

This is a process that aimed to identifying some issues related to data quality and we want to see the information of the data.

In [31]:
main_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 357 entries, 0 to 356
Data columns (total 40 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   property_id             357 non-null    object 
 1   unit_id                 357 non-null    object 
 2   average_baseline_price  357 non-null    float64
 3   bedroom                 357 non-null    int64  
 4   bathroom                124 non-null    float64
 5   beds                    124 non-null    float64
 6   capacity                117 non-null    float64
 7   ac                      123 non-null    float64
 8   balcony                 123 non-null    float64
 9   beachfront              123 non-null    float64
 10  breakfast               123 non-null    float64
 11  building_staff          123 non-null    float64
 12  cable_tv                123 non-null    float64
 13  essentials              123 non-null    float64
 14  garden                  123 non-null    fl

We see that the data types from main dataset is correct, so we don't need to change the datatypes. Next, we will access info from areacodes dataset using info() function.

In [32]:
areacodes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54 entries, 0 to 53
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   area_id                   54 non-null     object 
 1   area_name                 54 non-null     object 
 2   area_squad                53 non-null     float64
 3   employee_id               22 non-null     object 
 4   created_at                33 non-null     object 
 5   updated_at                30 non-null     object 
 6   deleted_at                0 non-null      float64
 7   temp_column               0 non-null      float64
 8   area_distance_to_airport  18 non-null     object 
 9   airport_pickup_price_idr  20 non-null     float64
 10  area_airport_name         20 non-null     object 
 11  region_id                 42 non-null     float64
dtypes: float64(5), object(7)
memory usage: 5.2+ KB


From information above, we conclude that we will change the dtypes of deleted_at and temp_column to object dtype and also we want to convert the area_distance_to_airport datatype to float64. But if we concern about data info, we realize that there is no data in deleted_at and temp_column, so we are going to drop these columns from areacodes_df


In [33]:
columns = ["deleted_at","temp_column"]
areacodes_df = areacodes_df.drop(columns,axis=1)

Next we will check it again using info() function

In [34]:
areacodes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54 entries, 0 to 53
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   area_id                   54 non-null     object 
 1   area_name                 54 non-null     object 
 2   area_squad                53 non-null     float64
 3   employee_id               22 non-null     object 
 4   created_at                33 non-null     object 
 5   updated_at                30 non-null     object 
 6   area_distance_to_airport  18 non-null     object 
 7   airport_pickup_price_idr  20 non-null     float64
 8   area_airport_name         20 non-null     object 
 9   region_id                 42 non-null     float64
dtypes: float64(3), object(7)
memory usage: 4.3+ KB


We have successfully deleted some columns from dataset and next we will change the datatype of area_distance_to_airport into float64 dtype and before convert dtype we can make function to change the data into minutes.

In [45]:
def convert_time_to_minutes(time_str):
    if isinstance(time_str, str):
        if 'minute' in time_str:
            return float(time_str.split()[0])
        elif 'hour' in time_str:
            return float(time_str.split()[0]) * 60
        else:
            raise ValueError(f"Unknown time format: {time_str}")
    elif isinstance(time_str, float) and np.isnan(time_str):
        return np.nan
    else:
        raise TypeError(f"Expected string or NaN, got {type(time_str)}")

Next we can implement that function.

In [46]:
column = ["area_distance_to_airport"]

for c in column:
  areacodes_df[c] =  areacodes_df[c].apply(convert_time_to_minutes).astype('float64')

Next we can check the data info and data conversion

In [47]:
areacodes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54 entries, 0 to 53
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   area_id                   54 non-null     object 
 1   area_name                 54 non-null     object 
 2   area_squad                53 non-null     float64
 3   employee_id               22 non-null     object 
 4   created_at                33 non-null     object 
 5   updated_at                30 non-null     object 
 6   area_distance_to_airport  18 non-null     float64
 7   airport_pickup_price_idr  20 non-null     float64
 8   area_airport_name         20 non-null     object 
 9   region_id                 42 non-null     float64
dtypes: float64(4), object(6)
memory usage: 4.3+ KB


In [48]:
areacodes_df.head(54)

area_id          area_name  area_squad employee_id        created_at  \
0    A0000          Undefined         NaN         NaN               NaN   
1    A0001         Yogyakarta         3.0       E0028               NaN   
2    A0002            Uluwatu         2.0       E0015               NaN   
3    A0003             Canggu         2.0       E0014               NaN   
4    A0004               Ubud         2.0       E0042               NaN   
5    A0005            Jakarta         3.0       E0001               NaN   
6    A0006            Bandung         3.0       E0001               NaN   
7    A0007             Phuket         2.0       E0042               NaN   
8    A0008             Bingin         1.0       E0036               NaN   
9    A0009           Balangan         1.0       E0004               NaN   
10   A0010           Jimbaran         1.0       E0013               NaN   
11   A0011            Ungasan         1.0       E0013               NaN   
12   A0012           Nusa Dua         1.0       E0013               NaN   
13   A0013      Padang Padang         2.0       E0037               NaN   
14   A0014     Gili Trawangan         2.0       E0042               NaN   
15   A0015           Seminyak         2.0       E0014               NaN   
16   A0016               Kuta         2.0       E0014               NaN   
17   A0017          Singapore         2.0       E0042               NaN   
18   A0018             Cemagi         2.0       E0014               NaN   
19   A0019             Penang         2.0       E0042               NaN   
20   A0020        Nusa Penida         4.0       E0017               NaN   
21   A0021           Gili Air         2.0       E0042  2018-07-02 16:51   
22   A0022              Sanur         2.0         NaN  2019-09-30 08:36   
23   A0023               test         2.0      345232  2020-03-08 09:17   
24   A0024            Tabanan         2.0         NaN  2021-05-18 14:11   
25   A0025          Singaraja         2.0         NaN  2021-10-06 02:13   
26   A0026             Dalung         0.0         NaN  2022-04-12 12:22   
27   A0027          Kintamani         0.0         NaN  2022-06-24 12:22   
28   A0028         Karangasem         5.0         NaN  2022-09-22 01:55   
29   A0029             Mengwi         2.0         NaN  2022-10-07 09:16   
30   A0030             Berawa         2.0         NaN  2022-10-07 09:16   
31   A0031          Pererenan         2.0         NaN  2022-10-07 09:16   
32   A0032             Umalas         2.0         NaN  2022-10-07 09:18   
33   A0033             Greece         0.0         NaN  2023-04-07 05:16   
34   A0034  Ngurah Rai Aiport         0.0         NaN  2023-04-07 05:16   
35   A0035              Kutuh         1.0         NaN  2023-04-27 06:11   
36   A0036      Nusa Ceningan         0.0         NaN  2023-06-08 08:39   
37   A0037             Pecatu         0.0         NaN  2023-07-25 06:15   
38   A0038         Sinduharjo         0.0         NaN  2023-09-02 03:23   
39   A0039         Tamantirto         0.0         NaN  2023-09-02 03:23   
40   A0040        Gunungketur         0.0         NaN  2023-09-02 03:23   
41   A0041         Panembahan         0.0         NaN  2023-09-02 03:23   
42   A0042          Sariharjo         0.0         NaN  2023-09-02 03:23   
43   A0043           Ngupasan         0.0         NaN  2023-09-02 03:23   
44   A0044        Banguntapan         0.0         NaN  2023-09-02 03:23   
45   A0045        Timbulharjo         0.0         NaN  2023-09-02 03:23   
46   A0046             Baciro         0.0         NaN  2023-09-02 03:23   
47   A0047        Maguwoharjo         0.0         NaN  2023-09-02 03:23   
48   A0049           Sinduadi         0.0         NaN  2023-09-02 03:23   
49   A0050       Caturtunggal         0.0         NaN  2023-09-02 03:23   
50   A0051            Bedugul         0.0         NaN  2023-09-07 03:13   
51   A0052          Kerobokan         0.0         NaN  2023-10-09 03:29   
52   A0053            Ketewe

Next we will check is there any duplicates in main and areacodes dataframe.

In [49]:
print("Main data duplicate count: ",main_df.duplicated().sum())
print("Areacodes data duplicate count: ",areacodes_df.duplicated().sum())

Main data duplicate count:  0
Areacodes data duplicate count:  0


After checking, it can be determined that both datasets do not have duplicates. Next we will check null values from both datasets.

In [50]:
main_df.isna().sum()

property_id                 0
unit_id                     0
average_baseline_price      0
bedroom                     0
bathroom                  233
beds                      233
capacity                  240
ac                        234
balcony                   234
beachfront                234
breakfast                 234
building_staff            234
cable_tv                  234
essentials                234
garden                    234
gym                       234
hair_dryer                234
hanger                    234
heating                   234
hot_water                 234
kitchen                   234
linens                    234
lock                      234
luggage_drop_off          234
parking                   234
pool                      234
private_entrance          234
shampoo                   234
tv                        234
washer                    234
wifi                      234
workspace                 234
lat                       244
lng       

From the main dataset,we can see that many missing values are found.

In [51]:
areacodes_df.isna().sum()

area_id                      0
area_name                    0
area_squad                   1
employee_id                 32
created_at                  21
updated_at                  24
area_distance_to_airport    36
airport_pickup_price_idr    34
area_airport_name           34
region_id                   12
dtype: int64

From the areacodes dataset, it can be seen that the column with the most missing values is area_distance_to_airport, with a total of 36 missing values.